Reduce unsafeness in WakeLock.cpp and WakeLockManager.cpp

jelee53 opened the pull request:
#### c97aa796879a235964dbe932a7988ae432a7220f
<pre>
Reduce unsafeness in WakeLock.cpp and WakeLockManager.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=295877">https://bugs.webkit.org/show_bug.cgi?id=295877</a>
<a href="https://rdar.apple.com/155767745">rdar://155767745</a>

Reviewed by Chris Dumez.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
(WebCore::MainThreadPermissionObserver::MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::~MainThreadPermissionObserver):
* Source/WebCore/Modules/permissions/PermissionController.cpp:
(WebCore::PermissionController::singleton):
(WebCore::PermissionController::shared): Deleted.
(WebCore::PermissionController::protectedShared): Deleted.
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp:
(WebCore::WakeLockManager::WakeLockManager):
(WebCore::WakeLockManager::~WakeLockManager):
(WebCore::WakeLockManager::ref const):
(WebCore::WakeLockManager::deref const):
(WebCore::WakeLockManager::visibilityStateChanged):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWakeLockManager):
* Source/WebCore/dom/Document.h:

Canonical link: <a href="https://commits.webkit.org/297404@main">https://commits.webkit.org/297404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19edadc2f030b8bec8d8796b2ae76e8bf662e6b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111580 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31247 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117615 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61851 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84778 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114527 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25487 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100427 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65219 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18562 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61445 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94873 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18631 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120856 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38633 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28701 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93686 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39010 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93510 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38641 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16430 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34669 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17987 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38522 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44006 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38185 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41521 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39887 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/59b1c3846085390c1135fc4d5ee0e6f87b753f0d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47968

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d8f83ec7a1d7a12cadb5640703e83f6b588e4d74. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47968

rniwa commented on Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h:
```diff
@@ -56,7 +56,7 @@ class WakeLockManager final : public VisibilityChangeClient {
     void visibilityStateChanged() final;
 
-    Document& m_document;
+    const Ref<Document> m_document;
```
Yeah, this will be a leak.

cdumez commented on Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
```diff
@@ -67,8 +67,8 @@ void WakeLock::request(WakeLockType lockType, Ref<DeferredPromise>&& promise)
-    bool hasTransientActivation = document->window() && document->window()->hasTransientActivation();
-    PermissionController::shared().query(document->clientOrigin(), PermissionDescriptor { PermissionName::ScreenWakeLock }, *document->page(), PermissionQuerySource::Window, [this, protectedThis = Ref { *this }, document = Ref { *document }, hasTransientActivation, promise = WTFMove(promise), lockType](std::optional<PermissionState> permission) mutable {
+    bool hasTransientActivation = document->window() && document->protectedWindow()->hasTransientActivation();
+    PermissionController::protectedShared()->query(document->clientOrigin(), PermissionDescriptor { PermissionName::ScreenWakeLock }, *document->page(), PermissionQuerySource::Window, [this, protectedThis = Ref { *this }, document = Ref { *document }, hasTransientActivation, promise = WTFMove(promise), lockType](std::optional<PermissionState> permission) mutable {
```
Like the other PR, we should drop `protectedShared()` and rename `shared()` to `singleton()`.

jelee53 committed:
Reduce unsafeness in WakeLock.cpp and WakeLockManager.cpp
https://bugs.webkit.org/show_bug.cgi?id=295877
rdar://155767745

Reviewed by Chris Dumez.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebCore/Modules/permissions/MainThreadPermissionObserver.cpp:
(WebCore::MainThreadPermissionObserver::MainThreadPermissionObserver):
(WebCore::MainThreadPermissionObserver::~MainThreadPermissionObserver):
* Source/WebCore/Modules/permissions/PermissionController.cpp:
(WebCore::PermissionController::singleton):
(WebCore::PermissionController::shared): Deleted.
(WebCore::PermissionController::protectedShared): Deleted.
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
* Source/WebCore/Modules/screen-wake-lock/WakeLock.cpp:
(WebCore::WakeLock::request):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.cpp:
(WebCore::WakeLockManager::WakeLockManager):
(WebCore::WakeLockManager::~WakeLockManager):
(WebCore::WakeLockManager::ref const):
(WebCore::WakeLockManager::deref const):
(WebCore::WakeLockManager::visibilityStateChanged):
* Source/WebCore/Modules/screen-wake-lock/WakeLockManager.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedWakeLockManager):
* Source/WebCore/dom/Document.h:

Canonical link: https://commits.webkit.org/297404@main
